File: pwa/todo-app/frontend/src/viewmodels/TaskViewModel.ts
import Task from "../models/TaskModel";
import Service from "../services/Service";

export class TaskViewModel {
    private static instance: TaskViewModel;
    private service: Service;
    private taskMap: Map<string, Task>;

    private constructor() {
        this.service = new Service(process.env.REACT_APP_API_URL);
        this.taskMap = new Map();
    }

    static getIntance(): TaskViewModel {
        return this.instance ?? (this.instance = new TaskViewModel());
    }

    async createTask(task: Task): Promise<Task[]> {
        await this.service.POST(`/task`, task);
        this.taskMap.set(task._id, task);
        return Array.from(this.taskMap.values());
    }

    async removeTask(task: Task): Promise<Task[]> {
        await this.service.DELETE(`/task/${task._id}`);        
        this.taskMap.delete(task._id);
        return Array.from(this.taskMap.values());
    }

    async updateTask(task: Task): Promise<Task[]> {
        await this.service.PATCH(`/task`, task);
        this.taskMap.set(task._id, task);
        return Array.from(this.taskMap.values());
    }

    async getAllTasks(user_id: string): Promise<Task[]> {
        const result = await this.service.GET(`/task/${user_id}`);
        result.value.tasks?.forEach(task => {
            this.taskMap.set(task._id,task);
        })
        return Array.from(this.taskMap.values());
    }
}

const taskViewModel = () => {
    return TaskViewModel.getIntance();
}

export default taskViewModel;

File: flutter/todo_app/frontend/lib/viewmodels/TaskViewModel.dart
import 'dart:convert';

import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:todo_app/models/Task.dart';
import 'package:todo_app/services/backend.interface.dart';

class TaskViewModel {
  //Singleton
  static final TaskViewModel _taskViewModel = TaskViewModel._internal();
  final BackendInterface backendConnection =
      BackendInterface(dotenv.env['API_URL']! + '/task');

  factory TaskViewModel() {
    return _taskViewModel;
  }

  TaskViewModel._internal();

  Map<String, Map<String, Task>> _groupedTaskMap = {
    TaskState.TODO.value: {},
    TaskState.DOING.value: {},
    TaskState.DONE.value: {}
  };

  Future<void> initViewModel(String user_id) async {
    this._groupedTaskMap[TaskState.TODO.value]!.clear();
    this._groupedTaskMap[TaskState.DOING.value]!.clear();
    this._groupedTaskMap[TaskState.DONE.value]!.clear();
    var response = await backendConnection.GET(endpoint: '/' + user_id);
    List<Map<String, dynamic>>? jsonList =
        response.value != null ? List.from(response.value['tasks']) : null;
    jsonList?.forEach((element) {
      Task task = Task.fromJson(element);
      this._groupedTaskMap[task.state.value]!.putIfAbsent(task.id, () => task);
    });
  }

  List<Task> getAllTasksByState(TaskState state) {
    return _groupedTaskMap[state.value]!.values.toList();
  }

  void addTask(Task task) {
    backendConnection.POST(endpoint: '/', body: task.toJson());
    this._groupedTaskMap[task.state.value]!.putIfAbsent(task.id, () => task);
  }

  void addAllTasks(List<Task> tasks) {
    tasks.forEach((task) {
      this.addTask(task);
    });
  }

  void removeTask(Task task) {
    backendConnection.DELETE(endpoint: '/' + task.id);
    this._groupedTaskMap[task.state.value]!.remove(task.id);
  }

  void updateTask(Task task) {
    backendConnection.PATCH(endpoint: '/', body: task.toJson());
    this._groupedTaskMap[task.state.value]!.update(task.id, (value) => task);
  }

  void updateTaskState(Task task, TaskState oldTaskState) {
    backendConnection.PATCH(endpoint: '/', body: task.toJson());
    this._groupedTaskMap[oldTaskState.value]!.remove(task.id);
    this._groupedTaskMap[task.state.value]!.putIfAbsent(task.id, () => task);
  }
}
